modyn/models/gpt2/gpt2.py

```diff
@@ -0,0 +1,116 @@
+from transformers import AutoTokenizer, GPT2Config, GPT2LMHeadModel
+
+from modyn.models.coreset_methods_support import CoresetSupportingModule
+
+
+class Gpt2:
+    # pylint: disable-next=unused-argument
+    def __init__(self, hparams: Any, device: str, amp: bool) -> None:
+        self.model = Gpt2Modyn(hparams)
+        self.model.to(device)
+
+
+"""
+Adapted from an example implementation of a GPT-2 model.
+This implementation uses the GPT-2 tokenizer from Hugging Face's Transformers library:
+https://huggingface.co/docs/transformers/model_doc/gpt2
+"""
+
+
+class Gpt2Modyn(CoresetSupportingModule):
+    def __init__(self, hparams: Any) -> None:
+        super().__init__()
+
+        # Use hparams to decide the GPT-2 version
+        model_name = hparams.model_name_or_path if hasattr(hparams, "model_name_or_path") else "gpt2-large"
+
+        # Assert that the model name is valid
+        valid_model_names = {"gpt2", "gpt2-medium", "gpt2-large", "gpt2-xl"}
+        assert model_name in valid_model_names, f"Invalid model name: {model_name}. Must be one of {valid_model_names}."
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2-large")
+        # self.tokenizer.pad_token = self.tokenizer.eos_token
+        # Load the specified GPT-2 model
+
+        self.tokenizer.add_special_tokens(
+            {
+                "eos_token": "</s>",
+                "bos_token": "<s>",
+                "unk_token": "<unk>",
+                "pad_token": "<pad>",
+                "mask_token": "<mask>",
+            }
+        )
+
+        self.tokenizer.padding_side = "right"
+        # Load the specified GPT-2 model
+        self.model = GPT2LMHeadModel.from_pretrained(model_name)
+        self.config = GPT2Config.from_pretrained(model_name)
+        # self.transformer = GPT2Model(self.config)
+
+    def forward(self, data: torch.Tensor, labels: torch.Tensor = None) -> torch.Tensor:
+        """Forward method for text generation or language modeling tasks.
+
+        Args:
+        - data (torch.Tensor): Tensor of shape (batch_size, seq_len, 2), where
+          the last dimension contains token IDs and attention masks.
+        - labels (torch.Tensor, optional): Tensor of labels for language modeling tasks.
+
+        Returns:
+        - output: The output logits or loss from the GPT-2 model.
+        """
+        # Split input into token IDs and attention masks
+        input_ids = data[:, :, 0]
+
+        attention_mask = data[:, :, 1]
+        # Forward pass through GPT-2
+
+        output = self.model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
+
+        return output
+
+    def get_last_layer(self) -> nn.Module:
+        """Retrieve the last layer (lm_head) of the model.
+
+        Returns:
+            The final linear layer of the GPT-2 model.
+        """
+        return self.model.lm_head
+
+    def freeze_params(self) -> None:
+        for par in self.model.parameters():
+            par.requires_grad = False
```

Comment:
Do we actually use this somewhere? It's a bit weird to have this because the parent class does not require this. I feel like in terms of abstractions this is not optimal.